components/ai_chat/browser/ai_chat_tab_helper.cc

```diff
@@ -0,0 +1,441 @@
+}
+
+const std::string& AIChatTabHelper::GetConversationHistoryString() {
+  std::vector<std::string> turn_strings;
+  for (const ConversationTurn& turn : chat_history_) {
+    turn_strings.push_back((turn.character_type == CharacterType::HUMAN
+                                ? ai_chat::kHumanPromptPlaceholder
+                                : ai_chat::kAIPromptPlaceholder) +
+                           turn.text);
+  }
+
+  history_text_ = base::JoinString(turn_strings, "");
+
+  return history_text_;
+}
+
+void AIChatTabHelper::AddToConversationHistory(const ConversationTurn& turn) {
+  chat_history_.push_back(turn);
+
+  for (auto& obs : observers_) {
+    obs.OnHistoryUpdate();
+  }
+}
+
+void AIChatTabHelper::UpdateOrCreateLastAssistantEntry(
+    const std::string& updated_text) {
+  if (chat_history_.empty() ||
+      chat_history_.back().character_type != CharacterType::ASSISTANT) {
+    AddToConversationHistory({CharacterType::ASSISTANT,
+                              ConversationTurnVisibility::VISIBLE,
+                              updated_text});
+  } else {
+    chat_history_.back().text = updated_text;
+  }
+
+  // Trigger an observer update to refresh the UI.
+  for (auto& obs : observers_) {
+    obs.OnHistoryUpdate();
+  }
+}
+
+void AIChatTabHelper::AddObserver(Observer* observer) {
+  observers_.AddObserver(observer);
+}
+
+void AIChatTabHelper::RemoveObserver(Observer* observer) {
+  observers_.RemoveObserver(observer);
+}
+
+void AIChatTabHelper::GeneratePageText() {
+  auto* primary_rfh = web_contents()->GetPrimaryMainFrame();
+
+  if (!primary_rfh) {
+    VLOG(1) << "Summary request submitted for a WebContents without a "
+               "primary main frame";
+    return;
+  }
+  FetchPageContent(web_contents(),
+                   base::BindOnce(&AIChatTabHelper::OnTabContentRetrieved,
+                                  weak_ptr_factory_.GetWeakPtr()));
+}
+
+void AIChatTabHelper::OnTabContentRetrieved(std::string contents_text,
+                                            bool is_video) {
+  if (contents_text.empty()) {
+    VLOG(1) << __func__ << ": No data";
+    return;
+  }
+  is_video_ = is_video;
+  article_text_ = contents_text;
+
+  // Prevent indirect prompt injections being sent to the AI model.
+  base::ReplaceSubstringsAfterOffset(&contents_text, 0, ai_chat::kHumanPrompt,
+                                     "");
+  base::ReplaceSubstringsAfterOffset(&contents_text, 0, ai_chat::kAIPrompt, "");
+  base::ReplaceSubstringsAfterOffset(&contents_text, 0, "<article>", "");
+  base::ReplaceSubstringsAfterOffset(&contents_text, 0, "</article>", "");
+  base::ReplaceSubstringsAfterOffset(&contents_text, 0, "<history>", "");
+  base::ReplaceSubstringsAfterOffset(&contents_text, 0, "</history>", "");
+  base::ReplaceSubstringsAfterOffset(&contents_text, 0, "<question>", "");
+  base::ReplaceSubstringsAfterOffset(&contents_text, 0, "</question>", "");
```

Comment:
Where did we get this indirect prompt injection sanitize list?

Reply:
Copy pasted from another place in the file. But it's based on the prompts in `ai_chat_prompts.grdp`. If those prompts contain "structured" data, then we include the structure here.